registry/credentials.go

```diff
@@ -123,3 +123,7 @@ func (cs Credentials) Merge(c Credentials) {
 		cs.m[k] = v
 	}
 }
+
+func (cs Credentials) String() string {
+	return fmt.Sprintf("{%v}", cs.m)
```

Comment:
What will this output if there's some credentials present?

Reply:
Something like:
```
"{map[localhost:5000:<registry creds for testuser@localhost:5000, from test>]}"
```
I've added a test to ensure this is the case, see also: `registry/credentials_test.go#TestStringShouldNotLeakPasswords`.
If you remove this implementation of `Credentials#String`, and let Go use the default one, you end up with something which does print the password:
```
"{map[localhost:5000:{testuser testpassword localhost:5000 test}]}"
```

Reply:
That does end up being rather chatty:

```
ts=2018-01-25T17:35:40.862006068Z caller=warming.go:117 \
 component=warmer \
 canonical_name=kube-registry.kube-system.svc.cluster.local:5000/mikeb/hello \
 creds="{map[kube-registry.kube-system.svc.cluster.local:5000:<registry creds for mikeb@kube-registry.kube-system.svc.cluster.local:5000, from hello:secret/kube-registry-docker-secret>]}" \
 err="Get https://kube-registry.kube-system.svc.cluster.local:5000/v2/: http: server gave HTTP response to HTTPS client"
```

(Note in particular the repetition of the image domain x 3)

Reply:
Well, since the original goal of this PR was to help investigate the issue that #897 fixed, I wanted it to be chatty to better see what was happening. 

Also, aren't the domains from the image and the credentials unrelated? (as in, coming from different sources?)

What would you rather have?

Reply:
>  I wanted it to be chatty to better see what was happening.

Yeah fair enough. Having full info in a trace is good for that; otherwise I think we'll mostly care only if there's a problem. But it's always a fine balance isn't it.

> Also, aren't the domains from the image and the credentials unrelated?

They are exactly the same. The key of a map entry is the domain given in the docker-config secret, which is also given to the `cred{}` value to report as its domain; the `cred{}` value to use for an image is picked by determining an image's canonical domain and indexing into the map at that key.

The problem fixed by #897 was that the domain of the authentication URL, rather than the domain of the canonical name of the image, was being used to determine which credentials to use. I noticed that was the case by observing that the value was `<zero creds>` rather than creds from the secret.

I think it'd be fine to log the map on errors or on entry to the method, and just use the creds specific to the image otherwise (i.e., in the client itself). The factory is an interstitial structure -- not much interesting goes on there anyway.

Reply:
Thanks a lot for the explanations and suggestions @squaremo.

> But it's always a fine balance isn't it.

Indeed.

> They are exactly the same.

OK, I'll drop `"canonical_name", id.CanonicalName()` then.

> log the map on errors

OK, I'll do that, i.e. if I understood well:
```go
logger.Log("auth", creds, "err", err)
```
> log the map [...] on entry to the method

I believe we [do almost this already](https://github.com/weaveworks/flux/blob/master/registry/client_factory.go#L79-L81) post #897, in the factory, when creating the client, though only when traces are enabled and only with the creds for the current domain, but that's probably enough?

> just use the creds specific to the image otherwise (i.e., in the client itself)

I guess that's covered by the previous point?

Reply:
I think where I've ended up is that your "fallback" suggestion -- log the canonical name, and creds, but only on error -- is the right thing. Sorry for the run-around!

Reply:
NB the test for not leaking passwords is certainly worth keeping whatever else happens.

Reply:
> log the canonical name, and creds, but only on error

OK, adding the canonical name back then :+1: